File: Intermediate/challenge_4_smart_home_devices/python/implementation.py
from typing import Protocol

# =========================
# INTERFACES (ISP)
# =========================

class Switchable(Protocol):
    """OOP: Interface (capability-based)
       SOLID: Interface Segregation Principle (ISP)
    """
    def turn_on(self) -> None: ...
    def turn_off(self) -> None: ...
    @property
    def is_on(self) -> bool: ...

class Adjustable(Protocol):
    """OOP: Interface (capability-based)
       SOLID: ISP — only devices that need adjustment implement this
    """
    def set_value(self, value: int) -> None: ...
    @property
    def value(self) -> int: ...

# =========================
# BEHAVIORS (SRP + COMPOSITION)
# =========================

class OnOffBehavior:
    """OOP: Encapsulated state + behavior
       SOLID: Single Responsibility Principle (SRP)
    """
    def __init__(self, initial_state: bool = False):
        self._state = initial_state

    def turn_on(self) -> None:
        self._state = True

    def turn_off(self) -> None:
        self._state = False

    @property
    def is_on(self) -> bool:
        return self._state

class ValueBehavior:
    """OOP: Encapsulation
       SOLID: SRP — manages numeric state & validation only
    """
    def __init__(self, value: int, min_value: int, max_value: int):
        if min_value > max_value:
            raise ValueError("Invalid value range")
        if not min_value <= value <= max_value:
            raise ValueError("Initial value out of range")

        self._value = value
        self._min = min_value
        self._max = max_value

    def set_value(self, value: int) -> None:
        if not self._min <= value <= self._max:
            raise ValueError("Value out of range")
        self._value = value

    @property
    def value(self) -> int:
        return self._value

# =========================
# DEVICES (COMPOSITION + DIP)
# =========================

class Light:
    """OOP: Concrete class
       SOLID:
       - Uses composition instead of inheritance
       - Depends on abstractions (DIP)
       - Open for extension (add behaviors), closed for modification (OCP)
    """
    def __init__(self, power: OnOffBehavior):
        self._power = power  # Composition

    def turn_on(self) -> None:
        self._power.turn_on()

    def turn_off(self) -> None:
        self._power.turn_off()

    @property
    def is_on(self) -> bool:
        return self._power.is_on

class Fan:
    """OOP: Concrete class
       SOLID:
       - ISP: Implements only needed capabilities
       - DIP: Behaviors injected
       - LSP: Can be used anywhere Switchable/Adjustable is expected
    """
    def __init__(self, power: OnOffBehavior, speed: ValueBehavior):
        self._power = power        # Composition
        self._speed = speed

    def turn_on(self) -> None:
        self._power.turn_on()

    def turn_off(self) -> None:
        self._power.turn_off()

    def set_value(self, value: int) -> None:
        if not self._power.is_on:
            raise RuntimeError("Fan must be on to change speed")
        self._speed.set_value(value)

    @property
    def is_on(self) -> bool:
        return self._power.is_on

    @property
    def value(self) -> int:
        return self._speed.value

# =========================
# USAGE (POLYMORPHISM)
# =========================

def toggle(device: Switchable) -> None:
    """OOP: Polymorphism
       SOLID: LSP — works with ANY Switchable
    """
    if device.is_on:
        device.turn_off()
    else:
        device.turn_on()


fan = Fan(
    power=OnOffBehavior(),
    speed=ValueBehavior(value=1, min_value=1, max_value=5)
)

light = Light(OnOffBehavior())

toggle(fan)
fan.set_value(3)

toggle(light)

File: Intermediate/challenge_4_smart_home_devices/cpp/implementation.cpp
#include <iostream>
#include <stdexcept>

// Interface for devices that can be switched on and off.
// Notes:
//  - Small interface representing the switchable capability.
//  - Pure virtual methods force concrete devices to implement on/off behavior,
//    while the virtual destructor makes polymorphic destruction safe.
class Switchable {
public:
    // Virtual destructor because Switchable is used polymorphically.
    virtual ~Switchable() = default;

    // Pure virtual methods: every concrete switchable device
    // must provide its own implementation.
    virtual void turn_on() = 0;
    virtual void turn_off() = 0;
};


// Separate interface for devices that support an adjustable value.
// This avoids forcing all switchable devices to implement set_value().
// Notes:
//  - Keeps set_value() out of Switchable so non-adjustable devices are not
//    forced to implement unused behavior. 
//  - int is passed by value because it is cheap to copy.
class Adjustable {
public:
    // Virtual destructor because Adjustable is used polymorphically.
    virtual ~Adjustable() = default;

    // Pure virtual method: adjustable devices must define
    // how their value is changed.
    virtual void set_value(int value) = 0;
};


// Light only supports switching on and off.
// It therefore implements only the Switchable interface.
// Notes:
//  - Light publicly implements the Switchable capability.
//  - is_on_ is initialized to false in-class.
//  - override verifies the inherited interface methods.
//  - The methods modify the Light's internal state.
class Light : public Switchable {
private:
    // In-class initialization: lights start turned off.
    bool is_on_{false};

public:
    void turn_on() override
    {
        is_on_ = true;
        std::cout << "Light turned on.\n";
    }

    void turn_off() override
    {
        is_on_ = false;
        std::cout << "Light turned off.\n";
    }
};


// Fan supports both switching and adjustable speed.
// Multiple inheritance is appropriate here because both base classes
// represent independent capabilities/interfaces.
// Notes:
//  - Fan implements two independent interfaces through multiple inheritance.
//  - Public inheritance allows it to be used as either Switchable or Adjustable.
//  - Members use in-class initialization.
//  - override verifies interface methods.
//  - set_value() validates input before changing the fan's internal state.
class Fan : public Switchable, public Adjustable {
private:
    // Fan starts turned off with speed 0.
    bool is_on_{false};
    int speed_{0};

public:
    void turn_on() override
    {
        is_on_ = true;
        std::cout << "Fan turned on.\n";
    }

    void turn_off() override
    {
        is_on_ = false;
        std::cout << "Fan turned off.\n";
    }

    void set_value(int speed) override
    {
        // Pass int by value because copying small numeric types is cheap.
        // Validate before modifying the object's state.
        if (speed < 0 || speed > 3) {
            throw std::invalid_argument(
                "Fan speed must be between 0 and 3."
            );
        }

        speed_ = speed;

        std::cout << "Fan speed set to "
                  << speed_
                  << ".\n";
    }
};


// Accept any device that implements Switchable.
// Notes:
//  - Pass by reference to avoid copying and preserve polymorphism.
//  - Reference is preferred over a pointer because a valid device is required
//    and nullptr handling is therefore unnecessary.
void activate(Switchable& device)
{
    device.turn_on();
}


// Accept only devices that implement Adjustable.
// A Light cannot be passed here because it does not implement Adjustable.
// Notes:
//  - Pass Adjustable by reference to avoid copying and preserve polymorphism.
//  - A reference is preferred because a valid adjustable device is required.
//  - Pass int by value because copying numeric types is cheap.
void adjust(Adjustable& device, int value)
{
    device.set_value(value);
}


int main()
{
    try {
        Light light;
        Fan fan;

        activate(light);
        activate(fan);

        adjust(fan, 2);

        light.turn_off();
        fan.turn_off();
    }
    catch (const std::exception& error) {
        std::cerr << "Error: "
                  << error.what()
                  << '\n';
    }

    return 0;
}
